Implement Duffing Spring Actuator.py

eh111eh opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
#26402 

#### Brief description of what is fixed or changed
The Duffing equation is a nonlinear second order differential equation used to model the dynamics of systems, damped and driven oscillators, with a cubic nonlinearity in the restoring force, which is given by $$\\ddot{x} + \delta\dot{x} + \beta x + \alpha x^3 = \gamma \cos(\omega t)\$$

However, as a full Duffing second order forced system will not work as an actuator, I modify the Duffing Spring to treat it as a non-second order system.
For $\beta>0$, the Duffing oscillator can be interpreted as a forced oscillator with a spring whose restoring force is written as $$F = -\beta x - \alpha x^3$$
Since the case  $\beta<0$ causes the chaotic motions, I will not consider this case for now. 

I Implemented a new class `DuffingSpring` to model the force exerted by a nonlinear spring, based on the Duffing equation: $F = -\beta x - \alpha x^3$
where $x$ is the displacement from the equilibrium position, $\beta$ is the linear spring constant, and $\alpha$ is the coefficient for the nonlinear cubic term.

#### Task
- [x] Add a unit test to verify the new class `DuffingSpring` works as expected. Refer to the existing unit tests for actuators as examples.
- [x] Demonstrate the force exerted by a nonlinear spring, based on the Duffing equation, as illustrated in the following example: https://pydy.readthedocs.io/en/latest/examples/mass-spring-damper.html

#### Other comments
Reference
- http://www.scholarpedia.org/article/Duffing_oscillator 
- https://www.cfm.brown.edu/people/dobrush/am34/Matlab/ch3/duffing.html
- https://www.cfm.brown.edu/people/dobrush/am34/Mathematica/ch3/duffing.html

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* physics.mechanics
  * implemented `DuffingSpring` to model the force exerted by a nonlinear spring, based on the Duffing equation
<!-- END RELEASE NOTES -->

eh111eh committed:
Update actuator.py

eh111eh committed:
Added tests

There are several issues with this test for the DuffingSpring class including
i. Type checking (attributes, etc),
ii. Force expression
iii. SimpifyError in test_to_loads
iv. Failed to raise error
I've addressed these issues to pass pytest

eh111eh committed:
Resolved assertion errors in `test_valid_constructor`

The following three assertion errors in test_valid_constructor have been addressed:
1. Resolved Issue: The initial assertion error was resolved by ensuring spring.linear_stiffness is now correctly instantiated as an ExprType. This was achieved by properly setting and converting the linear_stiffness attribute into a SymPy expression.
2. Resolved Issue: The second assertion error related to the equilibrium_length attribute mirrored the first and was similarly addressed.
3. Resolved Issue: The third assertion error, concerning the mismatch of the force attribute with its expected expression, was corrected by revising the force calculation implementation in the DuffingSpring class.

eh111eh committed:
Resolved test errors

The test error associated with test_properties_are_immutable has been resolved. There remains one last error to address: This error arises in the test_repr method of the TestDuffingSpring class during pytest collection. The problem is tied to the use of the equilibrium_length parameter, which, although specified in the pytest.mark.parametrize decorator, is missing from the method signature of test_repr. This mismatch results in the error message: "In test_repr: function uses no argument 'equilibrium_length'."

eh111eh committed:
Refining the DuffingSpring Class: Updates and Testing

changes in actuator.py:
1. The DuffingSpring class was modified to use property decorators, which provide a getter method for linear_stiffness, nonlinear_stiffness, pathway, and equilibrium_length. This change encapsulates the properties and prevents them from being modified after the object is created, enforcing immutability of the spring parameters.

2. The constructor (__init__) uses a try-except block to catch SympifyError, ensuring that inputs can be converted to SymPy expressions. If conversion fails, it raises an error informing the user.

changes in test_actuator.py
1. The pytest.mark.parametrize decorator is used to run the same test function with different sets of arguments, allowing comprehensive testing of the DuffingSpring constructor with various stiffness and equilibrium length parameters.

2. Tests were added to confirm that the DuffingSpring object initializes with the expected attributes and that these attributes match the parameters passed to the constructor.

3. Tests ensure that once a DuffingSpring object is created, its key properties cannot be altered, testing the enforcement of immutability.

4. The test_to_loads function calculates the expected forces based on the Duffing equation and compares them to the forces returned by the to_loads method of the DuffingSpring object.

5. The __repr__ method is tested to confirm it returns the expected string representation of the DuffingSpring instance.

Now it passes pytest

eh111eh committed:
Update test_actuator.py

eh111eh committed:
Update test_actuator.py

eh111eh committed:
Update test_actuator.py

eh111eh commented:
Hi, I've added a test unit for the DuffingSpring class in test_actuator.py and made minor modifications to the DuffingSpring class itself. The test unit covers the following functionalities:

- Fixture Setup
- Inheritance Checks
- Parametrized Tests
- Constructor Validity
- Invalid Constructor Arguments
- Immutability of Properties
- String Representation
- Force Calculation

All tests pass in pytest. I would greatly appreciate any feedback or comments on this work. Thank you!

tjstienstra commented on sympy/physics/mechanics/actuator.py:
```diff
@@ -873,3 +875,82 @@ def __repr__(self):
+
+class DuffingSpring(ForceActuator):
+    """A nonlinear spring based on the Duffing equation.
+    Explanation
+    ===========
+    Here, ``DuffingSpring`` represents the force exerted by a nonlinear spring based on the Duffing equation:
+    F = -beta*x-alpha*x**3, where x is the displacement from the equilibrium position, beta is the linear spring constant,
+    and alpha is the coefficient for the nonlinear cubic term.
+    Parameters
+    ==========
```
Could you add blank line above the headings `Explanation` and `Parameters`.

tjstienstra commented on sympy/physics/mechanics/actuator.py:
```diff
@@ -873,3 +875,82 @@ def __repr__(self):
+    """
+
+    def __init__(self, linear_stiffness, nonlinear_stiffness, pathway, equilibrium_length=S.Zero):
+        try:
+            self._linear_stiffness = sympify(linear_stiffness)
+            self._nonlinear_stiffness = sympify(nonlinear_stiffness)
+            self._equilibrium_length = sympify(equilibrium_length)
+        except SympifyError:
+            raise SympifyError("Input arguments could not be sympified.")
```
I would recommend removing the try statement, as the error message of sympify should be clear enough.

tjstienstra commented on sympy/physics/mechanics/actuator.py:
```diff
@@ -873,3 +875,82 @@ def __repr__(self):
+    def linear_stiffness(self):
+        return self._linear_stiffness
+
+    @linear_stiffness.setter
+    def linear_stiffness(self, value):
+        raise AttributeError("Cannot modify linear_stiffness")
```
If you only specify a getter of a property, trying to set a value will already raise an AttributeError. Therefore, you can remove the setters that only raise an AttributeError, i.e.:
```python
def __init__(self, ...):
    self._linear_stiffness = sympify(linear_stiffness, strict=True)
    ...

@property
def linear_stiffness(self):
    return self._linear_stiffness
```
If you prefer having a separate setter to match the implementation of `LinearSpring`, `LinearDamper`, etc., then I would propose trying to use the same format, i.e.:
```python
def __init__(self, ...):
    self.linear_stiffness = linear_stiffness
    ...

@property
def linear_stiffness(self):
    return self._linear_stiffness

@linear_stiffness.setter
def linear_stiffness(self, linear_stiffness):
        if hasattr(self, '_linear_stiffness'):
            msg = (
                f'Can\'t set attribute `linear_stiffness` to '
                f'{repr(linear_stiffness)} as it is immutable.'
            )
            raise AttributeError(msg)
        self._linear_stiffness = sympify(linear_stiffness, strict=True)
```

tjstienstra commented on sympy/physics/mechanics/actuator.py:
```diff
@@ -873,3 +875,82 @@ def __repr__(self):
+    @property
+    def force(self):
+        """The force produced by the Duffing spring."""
+        displacement = simplify(self.pathway.length - self.equilibrium_length)
```
Why do you want to simplify this expression?
In general, it is best to avoid `simplify` in the source code as complicated expressions may end up running forever. If I were to define a 'complex' pathway, e.g. two points that are defined w.r.t. each other with a big expression then this would cause a huge slowdown.

tjstienstra commented on sympy/physics/mechanics/tests/test_actuator.py:
```diff
@@ -660,3 +662,190 @@ def test_to_loads_with_reaction(self):
+        assert isinstance(spring.force, ExprType)
+        assert spring.force == force
+
+    @pytest.mark.parametrize('linear_stiffness', [NonSympifyable(), [NonSympifyable()]])
```
Why run the same test twice?

tjstienstra replied:
You could test `'k_linear'`, `None` as in `TestLinearSpring`. This would require to use `sympify(..., strict=True)` when setting the `linear_stiffness`.

tjstienstra reviewed (changes requested):
I am not familiar with the Duffing spring, but the implementation looks mostly good to me. Just suggested a few minor changes.

tjstienstra commented:
@eh111eh don't forget to add yourself to the mailmap.

eh111eh committed:
Update DuffingSpring and its test

eh111eh commented:
> I am not familiar with the Duffing spring, but the implementation looks mostly good to me. Just suggested a few minor changes.

Thank you for your thorough feedback. I have made the following changes:

- Added a blank line before and after `Explanation` and `Parameters`.
- Removed the try statement.
- Modified the setters to match the implementation in other classes, such as `LinearSpring`.
- Removed `simplify` from the displacement calculation. I originally applied `simplify` to resolve an issue where the actual force output did not match the expected force output and forgot to revert it after resolving it.
- Avoided repeating tests in `test_actuator.py` and added tests for `None`.
- Added my name and email to the mailmap.

I would appreciate your comments and feedback. Thank you!

eh111eh committed:
Update actuator.py

tjstienstra commented on sympy/physics/mechanics/actuator.py:
```diff
@@ -873,3 +875,93 @@ def __repr__(self):
+    """
+
+    def __init__(self, linear_stiffness, nonlinear_stiffness, pathway, equilibrium_length=S.Zero):
+        self._linear_stiffness = sympify(linear_stiffness, strict=True)
```
You could use `self.linear_stiffness = linear_stiffness`, as you've added the check in the setter.

eh111eh replied:
I have just updated the class!

eh111eh committed:
Update actuator.py

eh111eh committed:
Update actuator.py

tjstienstra reviewed (approved):
LGTM
